File: Android/java/app/src/main/java/org/oreilly/nmd/BrowseBooksAdapter.java
package org.oreilly.nmd;

import android.content.Intent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.gson.Gson;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class BrowseBooksAdapter extends RecyclerView.Adapter<BrowseBooksAdapter.ViewHolder> {

  private final BookDataSource mSource;

  public BrowseBooksAdapter(BookDataSource source) {
    super();
    mSource = source;
  }

  @Override
  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return new ViewHolder(new TextView(parent.getContext()));
  }

  @Override
  public void onBindViewHolder(ViewHolder holder, int position) {
    Book book = mSource.get(position);
    holder.mTextView.setTag(book);
    holder.mTextView.setText(book.getTitle());
  }

  @Override
  public int getItemCount() {
    return mSource.size();
  }

  public static class ViewHolder extends RecyclerView.ViewHolder {
    private TextView mTextView;
    public ViewHolder(@NonNull View itemView) {
      super(itemView);
      mTextView = (TextView) itemView;
      mTextView.setOnClickListener(this::showBook);
    }
    private void showBook(View view) {
      Book book = (Book) view.getTag();
      String json = new Gson().toJson(book);
      Intent intent = new Intent(view.getContext(), BookDetailActivity.class);
      intent.putExtra(BookDetailActivity.BOOK_JSON, json);
      view.getContext().startActivity(intent);
    }
  }


}

File: Android/kotlin/app/src/main/java/org/oreilly/nmd/BrowseBooksAdapter.kt
package org.oreilly.nmd

import android.content.Intent
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson

class BrowseBooksAdapter(private val source: BookDataSource) : RecyclerView.Adapter<BrowseBooksAdapter.ViewHolder>() {

    override fun getItemCount(): Int {
        return source.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        Log.d("MyApp", "creating a new textviw")
        return ViewHolder(TextView(parent.context))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val book = source[position]
        holder.textView.tag = book
        holder.textView.text = book.title
        Log.d("MyApp", "showing ${book.title}")
    }

    open class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textView: TextView = itemView as TextView

        init {
            textView.setOnClickListener { showBook(it) }
        }

        private fun showBook(view: View) {
            val book = view.tag as Book
            val json = Gson().toJson(book)
            val intent = Intent(view.context, BookDetailActivity::class.java)
            intent.putExtra(BookDetailActivity.BOOK_JSON, json)
            view.context.startActivity(intent)
        }
    }

}